File: java/scon-core/src/main/java/io/github/realmlabs/scon/SourceFormatter.java
package io.github.realmlabs.scon;

final class SourceFormatter {
    private SourceFormatter() {}

    static String formatSource(String source) {
        var document = Parser.parseDocument(source, null);
        var out = new StringBuilder();
        for (var token : Lexer.lex(source)) {
            if (token.kind().equals("comment")) out.append(token.text()).append('\n');
        }
        writeObjectBody(out, document.root(), 0);
        out.append('\n');
        return out.toString();
    }

    private static void writeObjectBody(StringBuilder out, Ast.ObjectNode object, int indent) {
        for (var member : object.members()) {
            out.append(" ".repeat(indent));
            if (member instanceof Ast.Include include) {
                out.append("include ").append(include.path().raw());
            } else if (member instanceof Ast.ObjectSpread spread) {
                out.append("...");
                writeSubstitution(out, spread.sub());
            } else if (member instanceof Ast.Field field) {
                writePath(out, field.path());
                out.append(" = ");
                writeValue(out, field.value(), indent);
            }
            out.append('\n');
        }
    }

    private static void writeValue(StringBuilder out, Ast.ValueNode value, int indent) {
        if (value instanceof Ast.NullNode) {
            out.append("null");
        } else if (value instanceof Ast.BoolNode bool) {
            out.append(bool.value() ? "true" : "false");
        } else if (value instanceof Ast.NumberNode number) {
            out.append(number.raw());
        } else if (value instanceof Ast.StringNode string) {
            out.append(string.raw());
        } else if (value instanceof Ast.SubstitutionNode substitution) {
            writeSubstitution(out, substitution);
        } else if (value instanceof Ast.ArrayNode array) {
            writeArray(out, array, indent);
        } else if (value instanceof Ast.ObjectValueNode object) {
            if (object.object().members().isEmpty()) {
                out.append("{}");
            } else {
                out.append("{\n");
                writeObjectBody(out, object.object(), indent + 2);
                out.append(" ".repeat(indent)).append('}');
            }
        }
    }

    private static void writeArray(StringBuilder out, Ast.ArrayNode array, int indent) {
        if (array.items().isEmpty()) {
            out.append("[]");
            return;
        }
        out.append("[\n");
        for (var item : array.items()) {
            out.append(" ".repeat(indent + 2));
            if (item instanceof Ast.ArrayValue value) {
                writeValue(out, value.value(), indent + 2);
            } else if (item instanceof Ast.ArraySpread spread) {
                out.append("...");
                writeSubstitution(out, spread.sub());
            }
            out.append(",\n");
        }
        out.append(" ".repeat(indent)).append(']');
    }

    private static void writeSubstitution(StringBuilder out, Ast.SubstitutionNode substitution) {
        out.append("${");
        writePath(out, substitution.path());
        out.append('}');
    }

    private static void writePath(StringBuilder out, Ast.PathNode path) {
        for (int i = 0; i < path.segments().size(); i++) {
            if (i > 0) out.append('.');
            var segment = path.segments().get(i);
            if (segment.quoted()) quote(out, segment.value());
            else out.append(segment.value());
        }
    }

    private static void quote(StringBuilder out, String value) {
        out.append('"');
        for (int i = 0; i < value.length(); i++) {
            char ch = value.charAt(i);
            switch (ch) {
                case '"' -> out.append("\\\"");
                case '\\' -> out.append("\\\\");
                case '\n' -> out.append("\\n");
                case '\r' -> out.append("\\r");
                case '\t' -> out.append("\\t");
                case '\b' -> out.append("\\b");
                case '\f' -> out.append("\\f");
                default -> out.append(ch);
            }
        }
        out.append('"');
    }
}

File: kotlin/scon-core/src/main/kotlin/io/github/realmlabs/scon/SourceFormatter.kt
package io.github.realmlabs.scon

internal class SourceFormatter(
    private val source: String,
    private val options: SconFormatOptions,
) {
    private data class FormatComment(
        val line: Int,
        val text: String,
        val span: SourceSpan,
        var emitted: Boolean = false,
    )

    private val lineIndex = LineIndex(source)
    private val comments = Lexer(source, "<format>").lex()
        .filter { it.kind == SconTokenKind.Comment }
        .map {
            FormatComment(
                line = lineIndex.sourcePosition(it.span.start).line,
                text = source.substring(it.span.start, it.span.end),
                span = it.span,
            )
        }

    fun format(document: ParsedDocument): String =
        buildString {
            emitObjectBody(document.root, 0, this)
            emitCommentsBefore(document.root.span.end, 0, this)
            if (!endsWith('\n')) append('\n')
        }

    private fun emitObjectBody(obj: AstObject, indent: Int, out: StringBuilder) {
        val members = obj.members.sortedBy { it.span.start }
        for ((index, member) in members.withIndex()) {
            val nextMemberStart = members.getOrNull(index + 1)?.span?.start ?: obj.span.end
            emitCommentsBefore(member.span.start, indent, out)
            when (member) {
                is AstField -> emitField(member, indent, out, nextMemberStart)
                is AstInclude -> {
                    writeIndent(out, indent)
                    out.append("include ")
                    out.append(formatStringLiteral(member.path.value))
                    emitInlineComment(member.span, out)
                    out.append('\n')
                }
                is AstObjectSpread -> {
                    writeIndent(out, indent)
                    out.append("...")
                    out.append(formatSubstitution(member.substitution.path))
                    emitInlineComment(member.span, out)
                    out.append('\n')
                }
            }
        }
        emitCommentsBefore(obj.span.end, indent, out)
    }

    private fun emitField(field: AstField, indent: Int, out: StringBuilder, nextMemberStart: Int) {
        writeIndent(out, indent)
        out.append(formatPath(field.path))
        val objectValue = field.value as? AstObjectValue
        if (objectValue != null) {
            if (fieldUsesEquals(field)) {
                out.append(" = {")
            } else {
                out.append(" {")
            }
            emitInlineComment(
                SourceSpan(field.span.start, objectValue.span.start + 1),
                out,
                beforeOffset = objectValue.value.members.minOfOrNull { it.span.start } ?: objectValue.span.end,
            )
            out.append('\n')
            emitObjectBody(objectValue.value, indent + options.indent, out)
            writeIndent(out, indent)
            out.append('}')
            emitInlineComment(field.span, out)
            out.append('\n')
            return
        }
        out.append(" = ")
        emitValue(field.value, indent, out)
        emitInlineComment(field.span, out, beforeOffset = nextMemberStart)
        out.append('\n')
    }

    private fun emitValue(value: AstValue, indent: Int, out: StringBuilder) {
        when (value) {
            is AstObjectValue -> {
                out.append("{\n")
                emitObjectBody(value.value, indent + options.indent, out)
                writeIndent(out, indent)
                out.append('}')
            }
            is AstArray -> emitArray(value, indent, out)
            is AstSubstitution -> out.append(formatSubstitution(value.path))
            is AstString, is AstNumber, is AstBool, is AstNull -> out.append(spanText(value.span).trim())
        }
    }

    private fun emitArray(array: AstArray, indent: Int, out: StringBuilder) {
        if (array.items.isEmpty()) {
            out.append("[]")
            return
        }
        out.append("[\n")
        for ((index, item) in array.items.withIndex()) {
            emitCommentsBefore(item.span.start, indent + options.indent, out)
            writeIndent(out, indent + options.indent)
            when (item) {
                is AstArrayValueItem -> emitValue(item.value, indent + options.indent, out)
                is AstArraySpread -> {
                    out.append("...")
                    out.append(formatSubstitution(item.substitution.path))
                }
            }
            if (index + 1 != array.items.size) out.append(',')
            emitInlineComment(item.span, out)
            out.append('\n')
        }
        emitCommentsBefore(array.span.end, indent + options.indent, out)
        writeIndent(out, indent)
        out.append(']')
    }

    private fun emitCommentsBefore(offset: Int, indent: Int, out: StringBuilder) {
        for (comment in comments) {
            if (comment.emitted || comment.span.start >= offset) continue
            writeIndent(out, indent)
            out.append(comment.text.trim())
            out.append('\n')
            comment.emitted = true
        }
    }

    private fun emitInlineComment(
        span: SourceSpan,
        out: StringBuilder,
        beforeOffset: Int = Int.MAX_VALUE,
    ) {
        val line = lineIndex.sourcePosition(span.start).line
        val comment = comments.firstOrNull {
            !it.emitted &&
                it.line == line &&
                it.span.start >= span.start &&
                it.span.start >= span.end &&
                it.span.start < beforeOffset
        } ?: return
        out.append(' ')
        out.append(comment.text.trim())
        comment.emitted = true
    }

    private fun fieldUsesEquals(field: AstField): Boolean =
        source.substring(field.path.span.end, field.value.span.start).contains('=')

    private fun spanText(span: SourceSpan): String =
        source.substring(span.start.coerceIn(0, source.length), span.end.coerceIn(0, source.length))

    private fun writeIndent(out: StringBuilder, indent: Int) {
        repeat(indent) { out.append(' ') }
    }
}

internal fun formatSubstitution(path: AstPath): String =
    "\${${formatPath(path)}}"

internal fun formatPath(path: AstPath): String =
    path.segments.joinToString(".") { segment ->
        if (segment.value.isSconIdentifier()) segment.value else formatStringLiteral(segment.value)
    }

internal fun formatStringLiteral(text: String): String =
    buildString {
        append('"')
        for (ch in text) {
            when (ch) {
                '"' -> append("\\\"")
                '\\' -> append("\\\\")
                '\n' -> append("\\n")
                '\r' -> append("\\r")
                '\t' -> append("\\t")
                '\b' -> append("\\b")
                '\u000C' -> append("\\f")
                else -> {
                    if (ch.isISOControl()) {
                        append("\\u")
                        append(ch.code.toString(16).padStart(4, '0'))
                    } else {
                        append(ch)
                    }
                }
            }
        }
        append('"')
    }

internal fun String.isSconIdentifier(): Boolean {
    if (isEmpty()) return false
    val first = first()
    if (!first.isAsciiLetter() && first != '_') return false
    return drop(1).all { it.isAsciiLetter() || it.isDigit() || it == '_' || it == '-' }
}

internal fun Char.isAsciiLetter(): Boolean =
    this in 'a'..'z' || this in 'A'..'Z'
